### PR TITLE
Improve SQL parser; accept LATENESS column attribute

### PR DESCRIPTION
**Pull request by mihaibudiu**

### CODE REVIEW

File: docs/docs/sql/datetime.md [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # Date/time operations
 
 `TIME` literals have the form `TIME 'HH:MM:SS.FFF`, where the
-fractional part is optional.
+fractional part is optional, and can have between 0 and 9 digits.
```
Comment (gz) [inline docs/docs/sql/datetime.md:4]:
```suggestion
fractional part is mandatory, and can have between 0 and 9 digits.
```

**Review by gz — APPROVED** [review]

File: docs/docs/sql/datetime.md (hunk 1 above)
Comment (gz) [reply docs/docs/sql/datetime.md:4]: I can't send a timestamp without the fractional part set.
see also https://github.com/feldera/dbsp/pull/441 if we want to change this

Comment (mihaibudiu) [reply docs/docs/sql/datetime.md:4]: it is not mandatory, it is optional.

Comment (mihaibudiu) [reply docs/docs/sql/datetime.md:4]: when present it can have that many digits

Comment (gz) [reply docs/docs/sql/datetime.md:4]:
```
curl 'http://localhost:8080/v0/pipelines/0189b2c6-0135-7f9e-95fd-93f99592da61/ingress/INPUTS?format=csv' -X POST -H 'User-Agent: Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/116.0' -H 'Accept: application/json' -H 'Accept-Language: en-US,en;q=0.5' -H 'Accept-Encoding: gzip, deflate, br' -H 'Referer: http://localhost:3000/' -H 'content-type: text/csv' -H 'Origin: http://localhost:3000' -H 'Connection: keep-alive' -H 'Sec-Fetch-Dest: empty' -H 'Sec-Fetch-Mode: cors' -H 'Sec-Fetch-Site: same-site' --data-raw '21:54:30'   
{"message":"parse error on input endpoint 'api-ingress-INPUTS-72537238-a294-4719-b973-799ffe6d1aa4': failed to deserialize csv record 'ByteRecord([\"21:54:30\"])': invalid time string '21:54:30': premature end of input","error_code":"ParseError","details":{"backtrace":"   0: dbsp_adapters::transport::http::input::HttpInputEndpoint::eoi\n   1: actix_web::handler::handler_service::{{closure}}::{{closure}}\n   2: <actix_web::resource::Resource<T> as actix_web::service::HttpServiceFactory>::register::{{closure}}::{{closure}}\n   3: <actix_web::middleware::logger::LoggerResponse<S,B> as core::future::future::Future>::poll\n   4: actix_http::h1::dispatcher::InnerDispatcher<T,S,B,X,U>::poll_response\n   5: <actix_http::h1::dispatcher::Dispatcher<T,S,B,X,U> as core::future::future::Future>::poll\n   6: <actix_http::service::HttpServiceHandlerResponse<T,S,B,X,U> as core::future::future::Future>::poll\n   7: <actix_service::and_then::AndThenServiceResponse<A,B,Req> as core::future::future::Future>::poll\n   8: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut\n   9: tokio::runtime::task::core::Core<T,S>::poll\n  10: tokio::runtime::task::harness::Harness<T,S>::poll\n  11: tokio::task::local::LocalSet::tick\n  12: <tokio::task::local::RunUntil<T> as core::future::future::Future>::poll\n  13: tokio::runtime::scheduler::current_thread::Context::enter\n  14: tokio::runtime::context::scoped::Scoped<T>::set\n  15: tokio::runtime::context::set_scheduler\n  16: tokio::runtime::scheduler::current_thread::CoreGuard::block_on\n  17: tokio::runtime::context::runtime::enter_runtime\n  18: tokio::runtime::runtime::Runtime::block_on\n  19: std::sys_common::backtrace::__rust_begin_short_backtrace\n  20: core::ops::function::FnOnce::call_once{{vtable.shim}}\n  21: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once\n             at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/alloc/src/boxed.rs:1985:9\n      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once\n             at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/alloc/src/boxed.rs:1985:9\n      std::sys::unix::thread::Thread::new::thread_start\n             at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/sys/unix/thread.rs:108:17\n  22: start_thread\n             at ./nptl/pthread_create.c:442:8\n  23: clone3\n             at ./misc/../sysdeps/unix/sysv/linux/x86_64/clone3.S:81\n","endpoint_name":"api-ingress-INPUTS-72537238-a294-4719-b973-799ffe6d1aa4","error":"failed to deserialize csv record 'ByteRecord([\"21:54:30\"])': invalid time string '21:54:30': premature end of input"}}%                                                     ```

Comment (gz) [reply docs/docs/sql/datetime.md:4]: FWIW I'm happy if it's optional, but that's not how it currently is we need to adjust the code then

Comment (mihaibudiu) [reply docs/docs/sql/datetime.md:4]: It most certainly is optional if you write SQL time literals. That part is tested.

Comment (mihaibudiu) [reply docs/docs/sql/datetime.md:4]: BTW: the documentation says "literals".
Not I/O.
I think I/O has to be documented separately.

Comment (mihaibudiu) [reply docs/docs/sql/datetime.md:4]: So you should probably file separate bugs for serde (I/O).

Comment (gz) [reply docs/docs/sql/datetime.md:4]: We should make sure the I/O is consistent with the literals to avoid confusion, but yes we need tests for it too.

**Comment by mihaibudiu** [comment]
@ryzhyk this is the first piece of the puzzle for handling lateness. 
With the commits so far the information from the CREATE TABLE makes it to the SourceOperator.
Next steps:
- propagate it
- synthesize the "watermark" circuits
- hook up potential users, such as window computations

**Pull request merged**